File: app/src/db/adapters/tauri/assets.ts
import { bytesFromWire, bytesToWire, invoke } from "./invoke";
import type { DataPort } from "../../DataPort";
import { assetDraftSchema } from "../../schemas";
import type { AssetDraft, AssetMetaRow } from "../../types";

type AssetMethods = Pick<DataPort, "getAssetMeta" | "getAssetBytes" | "listAssetMeta" | "putAsset">;

// Zod at every boundary (docs/CodingStandards.md) — see
// tauri/configs.ts's own comment on this same pattern. Parsed before the
// `Bytes` wire conversion, since `assetDraftSchema` expects a real
// `Uint8Array`, not the wire-format `number[]`.
export const createAssetMethods = (): AssetMethods => ({
    getAssetMeta: (assetId) => invoke<AssetMetaRow | null>("get_asset_meta", { assetId }),

    getAssetBytes: async (assetId) => {
        const bytes = await invoke<number[] | null>("get_asset_bytes", { assetId });
        return bytes === null ? null : bytesFromWire(bytes);
    },

    listAssetMeta: (ownerId) => invoke<AssetMetaRow[]>("list_asset_meta", { ownerId }),

    putAsset: (draft: AssetDraft) => {
        const parsed = assetDraftSchema.parse(draft);
        return invoke<AssetMetaRow>("put_asset", {
            draft: { ...parsed, Bytes: bytesToWire(parsed.Bytes) },
        });
    },
});

File: app/src-tauri/src/commands/assets.rs
use tauri::State;

use crate::error::AppError;
use crate::state::{lock, AppState};
use crate::store;
use crate::store::dto::{AssetDraft, AssetMetaRow};

#[tauri::command]
pub fn get_asset_meta(
    state: State<'_, AppState>,
    asset_id: String,
) -> Result<Option<AssetMetaRow>, AppError> {
    let conn = lock(&state)?;
    store::get_asset_meta(&conn, &asset_id)
}

#[tauri::command]
pub fn get_asset_bytes(
    state: State<'_, AppState>,
    asset_id: String,
) -> Result<Option<Vec<u8>>, AppError> {
    let conn = lock(&state)?;
    store::get_asset_bytes(&conn, &asset_id)
}

#[tauri::command]
pub fn list_asset_meta(
    state: State<'_, AppState>,
    owner_id: String,
) -> Result<Vec<AssetMetaRow>, AppError> {
    let conn = lock(&state)?;
    store::list_asset_meta(&conn, &owner_id)
}

#[tauri::command]
pub fn put_asset(state: State<'_, AppState>, draft: AssetDraft) -> Result<AssetMetaRow, AppError> {
    let conn = lock(&state)?;
    store::put_asset(&conn, &draft)
}
